openmc/filter.py

```diff
@@ -0,0 +1,1810 @@
+from abc import ABCMeta
+from collections import OrderedDict
+from collections.abc import Iterable
+import copy
+import hashlib
+from itertools import product
+from numbers import Real, Integral
+from xml.etree import ElementTree as ET
+
+import numpy as np
+import pandas as pd
+
+import openmc
+import openmc.checkvalue as cv
+from .cell import Cell
+from .material import Material
+from .mixin import IDManagerMixin
+from .surface import Surface
+from .universe import Universe
+
+
+_FILTER_TYPES = (
+    'universe', 'material', 'cell', 'cellborn', 'surface', 'mesh', 'energy',
+    'energyout', 'mu', 'polar', 'azimuthal', 'distribcell', 'delayedgroup',
+    'energyfunction', 'cellfrom', 'legendre', 'spatiallegendre',
+    'sphericalharmonics', 'zernike', 'zernikeradial', 'particle'
+)
+
+_CURRENT_NAMES = (
+    'x-min out', 'x-min in', 'x-max out', 'x-max in',
+    'y-min out', 'y-min in', 'y-max out', 'y-max in',
+    'z-min out', 'z-min in', 'z-max out', 'z-max in'
+)
+
+_PARTICLES = {'neutron', 'photon', 'electron', 'positron'}
+
+
+class FilterMeta(ABCMeta):
+    def __new__(cls, name, bases, namespace, **kwargs):
+        # Check the class name.
+        if not name.endswith('Filter'):
+            raise ValueError("All filter class names must end with 'Filter'")
+
+        # Create a 'short_name' attribute that removes the 'Filter' suffix.
+        namespace['short_name'] = name[:-6]
+
+        # Subclass methods can sort of inherit the docstring of parent class
+        # methods.  If a function is defined without a docstring, most (all?)
+        # Python interpreters will search through the parent classes to see if
+        # there is a docstring for a function with the same name, and they will
+        # use that docstring.  However, Sphinx does not have that functionality.
+        # This chunk of code handles this docstring inheritance manually so that
+        # the autodocumentation will pick it up.
+        if name != 'Filter':
+            # Look for newly-defined functions that were also in Filter.
```

Comment:
Does this code need to live here? It seems this is a utility function that should be generally available to all base/meta classes of this library and can be added as a mix-in or simply through a utils class.